### PR TITLE
Breaking: Fixable disable directives (fixes #11815)

### LINKED ISSUE #11815: --report-unused-disable-directives should be autofixable
**The version of ESLint you are using.**
v5.16.0

**The problem you want to solve.**
I want to enable `--report-unused-disable-directives` on a large existing codebase, but it surfaces too many problems to be practical to fix manually. So I'd need to write a script. At that point, why not make the directive autofixable?

**Your take on the correct solution to problem.**
Run `eslint --fix --report-unused-disable-directives`

**Are you willing to submit a pull request to implement this change?**
Yes

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/master/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[x] Add something to the core
[ ] Other, please explain:

#### What changes did you make? (Give an overview)

Implements https://github.com/eslint/rfcs/pull/78 as described.

#### Is there anything you'd like reviewers to focus on?

~I'm not confident I got all the places to update docs about the addition to `fixTypes` and friends...~ Thanks!

~Would you like me to split out the added logic in `apply-disable-directives.js` into a new file and individually unit test it? I was dubious that re-searching through the source file for the previously retrieved disable directives is the right approach for performance, so I held off in the initial commit.~ I ended up storing a `parentComment` in `linter.js` > `createDisableDirectives`, to hold off on re-scanning.

### CODE REVIEW

File: lib/linter/apply-disable-directives.js
Comment (mdjermanovic) [inline lib/linter/apply-disable-directives.js:29]: This logic could be wrong in some cases:

```js
/* eslint semi:2 */

const code = `const x = 5
// comment in the code`; // eslint-disable-line semi

```

autofixed to:

```js
/* eslint semi:2 */

const code = `const x = 5

```

The logic for block comments might be also wrong if the comment is multiline and the line where it starts happen to have the same length as the comment.

File: lib/linter/apply-disable-directives.js
Comment (mdjermanovic) [inline lib/linter/apply-disable-directives.js:163]: When we are removing individual rules, we could accidentally remove all rules but leave the directive (that will then disallow all rules):

```js
/* eslint no-undef: 2, no-proto: 2, no-iterator: 2 */

// eslint-disable-next-line no-proto,, no-iterator
undef_foo();
``` 

autofixed to:

```js
/* eslint no-undef: 2, no-proto: 2, no-iterator: 2 */

// eslint-disable-next-line 
undef_foo();

```

This doesn't happen with 1 comma (`no-proto, no-iterator`) because the ranges are overlapping in that comma (one fix wants to remove `no-proto,`, the other `, no-iterator`), so only one fix gets applies in the first pass, while the second pass removes the whole comment because at that point there is only one rule in the directive.

The `,,` example is an edge case, but relying on the overlapping of successive ranges seems error-prone. Can we instead replace the whole comment with a new version that doesn't contain the rule?

Fox example, instead of a fix on a smaller range:

`no-proto,` -> ` `

the error for `no-proto` would produce a fix on the range of the whole comment:

`// eslint-disable-next-line no-proto,, no-iterator` -> `// eslint-disable-next-line , no-iterator`

**Review by mdjermanovic — CHANGES_REQUESTED** [review]

**Comment by aladdin-add** [comment]
friendly ping @JoshuaKGoldberg

**Comment by JoshuaKGoldberg** [comment]
Hey thanks for the reviews and the ping! I'm about halfway through resolving the feedback locally and hope to finish within this week :)

File: lib/linter/apply-disable-directives.js
Comment (JoshuaKGoldberg) [reply lib/linter/apply-disable-directives.js:99]: Great, it's a bit of a relief not to have to do all that logic! In the latest commits I've simplified it to always replace with a `" "`, to keep things simple.

**Comment by JoshuaKGoldberg** [comment]
Status update: I've added a "grouping" flow so that the logic knows whether an entire comment should be removed rather than only some of the rule IDs within. I ran out of time today to update the existing unit tests and hope to get to that soon, after which I'll touch up some of the mentioned edge cases around commas and spaces.

File: lib/linter/apply-disable-directives.js
Comment (aladdin-add) [inline lib/linter/apply-disable-directives.js:122]: lodash has been removed from deps in https://github.com/eslint/eslint/pull/14287. can you just use the `Array.flatMap()`?

Comment (JoshuaKGoldberg) [reply lib/linter/apply-disable-directives.js:122]: Exciting! `Array.flatMap()` is unavailable until support for Node 10 is dropped, but I can use the `flatMap` added in that PR once I merge in from `eslint:master`. 👍

**Comment by JoshuaKGoldberg** [comment]
@mdjermanovic I _think_ I've addressed all the edge cases you pointed out? It no longer attempts to get spacing perfectly right, and instead always plops a `" "` in as the fix.

Also, apologies in advance for not writing unit tests for the new edge cases: if this new approach looks good to you, is there a particular way you'd like me to do that? I'm thinking of pulling out `processUnusedDisableDirectives` into its own file so I can unit test that directly, but I can't tell if that goes against the spirit of the existing ESLint tests setup.

**Comment by TSMMark** [comment]
As an eslint user I'm really looking forward to this. I ran lint with `--report-unused-disable-directives` at work on our codebase and there were several hundred offenses. We're holding off on making those fixes manually, knowing this new feature is around the corner, and we'll soon be able to easily autofix everything for free! Thanks for your efforts @JoshuaKGoldberg and eslint core team!

**Comment by Brian-Bartels** [comment]
+1 to this fix being a huge benefit to my work. I recently had to manually remove 130 disables and the whole time I was wishing this existed

File: lib/linter/apply-disable-directives.js
Comment (mdjermanovic) [inline lib/linter/apply-disable-directives.js:64]:
```suggestion
    const commentStart = commentToken.range[0];
```

Nodes, tokens, and comments have a `range` property. We can use `commentToken.range[0]` instead of calculating it from `loc`.

Also, it looks like we don't have any tests where this function will be used?

File: lib/linter/apply-disable-directives.js
Comment (mdjermanovic) [inline lib/linter/apply-disable-directives.js:97]: Similar to previous, we already have this range as `commentToken.range`.

**Review by mdjermanovic — CHANGES_REQUESTED** [review]

**Review by mdjermanovic — CHANGES_REQUESTED** [review]

File: lib/linter/apply-disable-directives.js
Comment (mdjermanovic) [inline lib/linter/apply-disable-directives.js:84]: This could still find a wrong index, for example:

```js
/* eslint no-undef: 2, no-undefined: 2 */

/* eslint-disable-next-line no-undefined, no-undef */
undefined;

```

will be autofixed to:

```js
/* eslint no-undef: 2, no-undefined: 2 */

/* eslint-disable-next-line ined,  */
undefined;

```

(first pass removed `no-undef` from `no-undefined`, the second removed `no-undef`).

Could we maybe use regular expressions to more precisely locate the rule? It could be something like this:

```suggestion
function createIndividualDirectivesRemoval(directives, commentToken, sourceCode) {
    const listOffset = /^\s*\S+\s+/u.exec(commentToken.value)[0].length;
    const listText = commentToken.value
        .slice(listOffset) // remove eslint-*
        .split(/\s-{2,}\s/u)[0] // remove -- directive comment
        .trimRight();
    const listStart = commentToken.range[0] + 2 + listOffset;

    return directives.map(directive => {
        const { ruleId } = directive;
        const match = new RegExp(String.raw`(?:^|,)\s*${ruleId}\s*(?:$|,)`, "u").exec(listText);
        const ruleOffset = match.index;
        const ruleEndOffset = ruleOffset + match[0].length;
        const ruleText = listText.slice(ruleOffset, ruleEndOffset);

        return {
            description: ruleId,
            fix: {
                range: [
                    listStart + ruleOffset + (ruleText.startsWith(",") && ruleText.endsWith(",") ? 1 : 0),
                    listStart + ruleEndOffset
                ],
                text: ""
            },
            position: directive.unprocessedDirective
        };
    });
}
```

**Review by mdjermanovic — CHANGES_REQUESTED** [review]

**Comment by mdjermanovic** [comment]
> I'm not confident I got all the places to update docs about the addition to `fixTypes` and friends

API docs are good (no need to update the CLIEngine section as it will be removed soon).

CLI docs: [here](https://github.com/eslint/eslint/blob/master/docs/user-guide/command-line-interface.md#options) (under "Fixing problems"), [here](https://github.com/eslint/eslint/blob/master/docs/user-guide/command-line-interface.md#--fix-type), and [here](https://github.com/eslint/eslint/blob/e837d811e9275e43ccc266aba10e9c0833e0376c/lib/options.js#L154) (this is printed with `eslint -h`).

File: lib/shared/types.js
Comment (mdjermanovic) [inline lib/shared/types.js:128]: I think this is a typedef for `meta` properties in rule objects, so `rule.meta.type` cannot be `"directive"`.

File: lib/cli-engine/cli-engine.js
Comment (mdjermanovic) [inline lib/cli-engine/cli-engine.js:342]:
```suggestion
    return Boolean(rule && rule.meta && fixTypes.has(rule.meta.type));
```

**Comment by mdjermanovic** [comment]
> Also, apologies in advance for not writing unit tests for the new edge cases: if this new approach looks good to you, is there a particular way you'd like me to do that? 

Sorry, I missed this.

The approach with grouping errors into one error when all rules from the comment are unused LGTM. I was thinking about something different, but this looks better :+1: 

> I'm thinking of pulling out `processUnusedDisableDirectives` into its own file so I can unit test that directly, but I can't tell if that goes against the spirit of the existing ESLint tests setup.

We'd prefer unit tests for modules (`apply-disable-directives.js`) over tests for internal functions like `processUnusedDisableDirectives`.

It would be also good to add some higher-level tests, in `tests/lib/linter/linter.js`. Also, if it's easier to write tests in `linter.js` than unit tests for `apply-disable-directives.js`, it's perfectly fine that most test cases end up in `linter.js`.

File: lib/linter/apply-disable-directives.js
Comment (JoshuaKGoldberg) [reply lib/linter/apply-disable-directives.js:64]: Nice, that `range` property -> not needing `sourceCode.lineStartIndices` makes this a lot simpler! Thanks 😄

Comment (JoshuaKGoldberg) [reply lib/linter/apply-disable-directives.js]: Awesome, that `range` property completely gets rid of the need to pass around `sourceCode`. Thanks!

File: tests/lib/linter/linter.js
Comment (JoshuaKGoldberg) [inline tests/lib/linter/linter.js:3373]: Was there anything in particular that should be tested at this level? I put in some unit-level ones into `apply-disable-directives.js` around commas, spacing, and ordering of rule IDs.

**Review by btmills — COMMENTED** [review]
I’m still working my way through tests, but here’s a few quick fixes.

File: docs/user-guide/command-line-interface.md
Comment (mdjermanovic) [inline docs/user-guide/command-line-interface.md:50]: We should also update [`--fix-type` section](https://github.com/eslint/eslint/blob/master/docs/user-guide/command-line-interface.md#--fix-type) in the same document.

**Review by mdjermanovic — CHANGES_REQUESTED** [review]

File: lib/linter/apply-disable-directives.js
Comment (mdjermanovic) [inline lib/linter/apply-disable-directives.js:62]:
```suggestion
            description: `'${ruleId}'`,
```

**Comment by nzakas** [comment]
@JoshuaKGoldberg just checking in to see if you're still working on this?

**Comment by JoshuaKGoldberg** [comment]
Sorry, I am - I missed the notification for the most recent round of reviews. Thanks for the ping (again)! I'll have them fixed up by the end of this weekend.

File: tests/lib/linter/linter.js
Comment (JoshuaKGoldberg) [reply tests/lib/linter/linter.js:3374]: Marking as resolved since there's been a review since the comment.

File: lib/linter/apply-disable-directives.js
Comment (mdjermanovic) [inline lib/linter/apply-disable-directives.js:219]:
```suggestion
            ...options.disableFixes ? {} : { fix }
```

`null` isn't an allowed value of the `fix` property, the property should be omitted if no fix is available.

File: tests/lib/linter/apply-disable-directives.js
Comment (mdjermanovic) [inline tests/lib/linter/apply-disable-directives.js:701]:
```suggestion
```

**Review by mdjermanovic — CHANGES_REQUESTED** [review]
Looks great, I have only one suggestion about the `fix` property.

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks!

**Comment by TSMMark** [comment]
Tested this branch on a medium sized codebase, and it seems to work beautifully. So far I've only noticed one minor issue:

I don't know if this is in-scope, or if there's another lint rule that will report/fix this, but I noticed when eslint-disable rules are removed, there's a possibility to end up with multiple adjacent spaces in the disable comment.

Example:

Screen Shot 2021-08-03 at 3 24 05 PM

Thanks again for your work on this

**Comment by mdjermanovic** [comment]
@TSMMark thanks for testing, and glad it works well for you!

> I don't know if this is in-scope, or if there's another lint rule that will report/fix this, but I noticed when eslint-disable rules are removed, there's a possibility to end up with multiple adjacent spaces in the disable comment.
> 
> Example:
> 
> Screen Shot 2021-08-03 at 3 24 05 PM

It's always difficult to produce a fix with the right formatting. We're usually making a minimal change that fixes the reported problem, and then rely on other rules to fix the whitespace afterward. Unfortunately, in this case there is no such rule. Since this is a minor issue that doesn't affect the correctness of this feature, and the solution doesn't seem trivial, I wouldn't address it in this PR (which is already fairly complex). If you open a separate issue, someone might be able to tweak the fix in another PR.

**Pull request merged**
